kart/diff_structs.py

```diff
@@ -430,25 +433,117 @@ def __json__(self):
         return {k: v for k, v in self.items()}
 
 
+class InvalidatedDeltaDiff(Exception):
+    pass
+
+
 class DeltaDiff(Diff):
     """
     A DeltaDiff is the inner-most type of Diff, the one that actually contains Deltas.
     Since Deltas know the keys at which they should be stored, a DeltaDiff makes sure to store Deltas at these keys.
+
+    === Using DeltaDiff with an iterator of Deltas ===
+    It is possible to pass in an iterator of Deltas (e.g. a generator) to the DeltaDiff constructor,
+    in which case the Deltas are NOT immediately stored in the DeltaDiff.
+    This is useful because there may be a lot of Deltas, and we don't want to store them in memory.
+
+    The only correct way to consume a DeltaDiff populated by a generator is to call `iter_items()`,
+    which will consume the iterator as it yields Deltas.
+    Calling that method will invalidate the DeltaDiff, so it cannot be used again (doing so will throw an exception)
+
+    Calling any other dict-like method (keys(), items(), len() etc) will consume the iterator and store its
+    contents in memory, which may be expensive.
```

Comment:
Do we _need_ this as a "dict-like" object considering all the `self._evaluate_lazy_initial_contents()` calls in the methods you're adding?

Or would it be better to treat it a bit like Django's QuerySet/ValuesQuerySet or QueryDict/MutableQueryDict classes where you explicitly get a related-but-different class for the occasions when you need to do special things, and only that class implements the full interface?

Reply:
I had the same thought, but I didn't see a way to change it without being quite invasive. The problem is that sometimes we concatenate these diffs together, usually when you do a diff between a commit and a working copy. In that case we get a commit<>commit diff, and then we get a WC diff and we call `Diff.concatenate(diff1, diff2)` to merge them together.

In that situation we need to be able to access the diffs in a dicty way so we can merge the contents. This can't be done without consuming the generator function into memory.

So we could have a completely separate set of diff structs for commit<>WC diffs, but it would probably cause a lot of headaches IMO. vs this relatively isolated hack

Reply:
I think the reasoning/approach is good, I'm more worried about a footgun for accidentally starting to always consume the iterator when we make a seemingly innocuous change in the future.

> So we could have a completely separate set of diff structs for commit<>WC diffs

I didn't mean a completely separate set of structs, more like the existing lazy class having with a `.consume()`/`.resolve()`/etc method which consumes the iterator and returns/recasts to some sort of ResolvedDataDiff subclass, which has the `__len__`/`__getitem__`/`keys`/`values`/etc methods defined on it.

Reply:
See #1034 for a partial implementation of this idea